### PR TITLE
Fix Psalm-issues batch #1

### PR DESCRIPTION
**Pull request by tvdijen**
Fixes all Psalm-issues in all subdirs under `lib/SimpleSAML`.
I hope it is not too big to review. I've added a few comments to the parts that are likely to raise questions

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -43,7 +43,7 @@
         "ext-json": "*",
         "ext-mbstring": "*",
         "gettext/gettext": "^4.6",
-        "jaimeperez/twig-configurable-i18n": "^2.0",
+        "jaimeperez/twig-configurable-i18n": "^2.1",
```
Comment (tvdijen) [inline composer.json]: We have namespaced `jaimeperez/twig-configurable-i18n` to solve some issues regarding inheritance from actual Twig-classes. Psalm doesn't recognize inheritance when mixing namespaced / non-namespaced classes because class-aliases are only created at runtime and Psalm doesn't actually execute code.

File: lib/SimpleSAML/Locale/Translate.php [hunk 2]
```diff
@@ -108,7 +108,7 @@ private function getDictionary($name)
                 $fileName = substr($name, $sepPos + 1);
                 $dictDir = Module::getModuleDir($module).'/dictionaries/';
             } else {
-                $dictDir = $this->configuration->getPathValue('dictionarydir', 'dictionaries/');
+                $dictDir = $this->configuration->getPathValue('dictionarydir', 'dictionaries/') ?: 'dictionaries/';
```
Comment (tvdijen) [inline lib/SimpleSAML/Locale/Translate.php]: This may look weird, but if the configuration-setting is set to `null`, then Configuration::getPathValue will return `null` rather than the default value...

File: lib/SimpleSAML/XHTML/Template.php [hunk 3]
```diff
@@ -261,17 +264,20 @@ private function setupTwigTemplatepaths()
...
     {
         $auto_reload = $this->configuration->getBoolean('template.auto_reload', true);
         $cache = $this->configuration->getString('template.cache', false);
+
         // set up template paths
         $loader = $this->setupTwigTemplatepaths();
+
         // abort if twig template does not exist
         if (!$loader->exists($this->twig_template)) {
-            return false;
```
Comment (tvdijen) [inline lib/SimpleSAML/XHTML/Template.php]: This leads to Psalm-issues all over the place because now `$this->twig` can be false and we don't test for it..
It does however not make any sense to continue execution if the template-file does not exist, so raising an exception makes much more sense. This seems like a remnant from the days where we would fall back to php-templates if no Twig-template exists.

File: lib/SimpleSAML/Metadata/SAMLParser.php [hunk 4]
```diff
@@ -333,10 +334,6 @@ public static function parseDescriptorsFile($file)
             throw new \Exception('Failed to read XML from file: '.$file);
         }
 
-        if ($doc->documentElement === null) {
-            throw new \Exception('Opened file is not an XML document: '.$file);
```
Comment (jaimeperez) [inline lib/SimpleSAML/Metadata/SAMLParser.php]: Why is this removed? Can't `documentElement` be `null`?

File: lib/SimpleSAML/Metadata/SAMLParser.php [hunk 5]
```diff
@@ -1118,13 +1115,12 @@ private static function processExtensions($element, $parentExtensions = [])
                     $ret['UIInfo']['PrivacyStatementURL'] = $e->getPrivacyStatementURL();
 
                     foreach ($e->getKeywords() as $uiItem) {
-                        if (!($uiItem instanceof \SAML2\XML\mdui\Keywords)
-                            || ($uiItem->getKeywords() === [])
-                            || ($uiItem->getLanguage() === null)
-                        ) {
+                        $keywords = $uiItem->getKeywords();
```
Comment (jaimeperez) [inline lib/SimpleSAML/Metadata/SAMLParser.php]: Is `$uiItem` guaranteed here to be an instance of `\SAML2\XML\mdui\Keywords`?

File: lib/SimpleSAML/XHTML/Template.php (hunk 3 above)
Comment (jaimeperez) [reply lib/SimpleSAML/XHTML/Template.php]: This makes me realize of a possible issue. Is it still possible for modules using twig templates already to work, without `usenewui` being set to `true`?

File: lib/SimpleSAML/XML/Parser.php [hunk 6]
```diff
@@ -41,7 +41,10 @@ public static function fromSimpleXMLElement(\SimpleXMLElement $element)
         /* Create a new parser with the xml document where the namespace definitions
          * are added.
          */
-        $parser = new Parser($element->asXML());
+
+        /** @var string $xml */
+        $xml = $element->asXML();
+        $parser = new Parser($xml);
         return $parser;
```
Comment (jaimeperez) [inline lib/SimpleSAML/XML/Parser.php]: At some point, we end up with some very ugly code due to psalm 😞 

This should be `return new Parser($element->asXML());`...

File: lib/SimpleSAML/XML/Signer.php [hunk 7]
```diff
@@ -303,8 +303,10 @@ public function sign($node, $insertInto, $insertBefore = null)
             $options
         );
 
-        /** @var \RobRichards\XMLSecLibs\XMLSecurityKey $this->privateKey */
-        $objXMLSecDSig->sign($this->privateKey);
+        /** @var \RobRichards\XMLSecLibs\XMLSecurityKey $privateKey */
```
Comment (jaimeperez) [inline lib/SimpleSAML/XML/Signer.php]: Shouldn't this be fixed in the definition of the property, instead of here?

File: lib/SimpleSAML/Metadata/SAMLParser.php (hunk 4 above)
Comment (tvdijen) [reply lib/SimpleSAML/Metadata/SAMLParser.php]: Nope, it's always a DOMElement:
https://www.php.net/manual/en/class.domdocument.php#domdocument.props.documentelement

File: lib/SimpleSAML/Metadata/SAMLParser.php (hunk 5 above)
Comment (tvdijen) [reply lib/SimpleSAML/Metadata/SAMLParser.php]: Yes it is:
https://github.com/simplesamlphp/saml2/blob/release-3.x/src/SAML2/XML/mdui/UIInfo.php#L59
https://github.com/simplesamlphp/saml2/blob/release-3.x/src/SAML2/XML/mdui/UIInfo.php#L104

File: lib/SimpleSAML/Metadata/SAMLParser.php (hunk 4 above)
Comment (jaimeperez) [reply lib/SimpleSAML/Metadata/SAMLParser.php]: Great! 👍

File: lib/SimpleSAML/XML/Parser.php (hunk 6 above)
Comment (tvdijen) [reply lib/SimpleSAML/XML/Parser.php]: I may be able to fix that using a `/** @psalm-return Parser */` on the method's docblock.. I'll give it a try

File: lib/SimpleSAML/XML/Signer.php (hunk 7 above)
Comment (tvdijen) [reply lib/SimpleSAML/XML/Signer.php]: No, it can be `false`, but in this case we've already tested for that here:
https://github.com/simplesamlphp/simplesamlphp/blob/9f97a5dd7c353c4fde3efbf8786577ba69713382/lib/SimpleSAML/XML/Signer.php#L286

I think I can clean this up.. I'll give it a shot

File: lib/SimpleSAML/Metadata/SAMLParser.php (hunk 5 above)
Comment (jaimeperez) [reply lib/SimpleSAML/Metadata/SAMLParser.php]: In theory, it's not, as you can set the `Keywords` property there to an array of whatever objects. Look at [this](https://github.com/simplesamlphp/saml2/blob/release-3.x/src/SAML2/XML/mdui/UIInfo.php#L112-L120). However, in this case I think it's safe, since the property would only get populated with `\SAML2\XML\mdui\Keywords` objects via the constructor 👍

File: lib/SimpleSAML/XML/Signer.php (hunk 7 above)
Comment (jaimeperez) [reply lib/SimpleSAML/XML/Signer.php]: Ah, I see! Thanks for clarifying, it's really difficult to review this without looking at the entire code...

File: lib/SimpleSAML/XML/Parser.php (hunk 6 above)
Comment (jaimeperez) [reply lib/SimpleSAML/XML/Parser.php]: Good one! I was completely unaware of that option...

File: lib/SimpleSAML/XML/Parser.php [hunk 8]
```diff
@@ -41,11 +42,7 @@ public static function fromSimpleXMLElement(\SimpleXMLElement $element)
         /* Create a new parser with the xml document where the namespace definitions
          * are added.
          */
-
-        /** @var string $xml */
-        $xml = $element->asXML();
-        $parser = new Parser($xml);
-        return $parser;
+        return new Parser($xml = $element->asXML());
```
Comment (jaimeperez) [inline lib/SimpleSAML/XML/Parser.php]: The problem here is not the return value, but the argument passed to `\SimpleSAML\XML\Parser::__construct()`, which cannot be `false`. `asXML()` can actually return false:

https://www.php.net/manual/en/simplexmlelement.asxml.php

Comment (jaimeperez) [reply lib/SimpleSAML/XML/Parser.php]: I wonder if we should actually check for error. I'm trying to see under what circumstances `false` can be returned by `asXML()`, but I'm failing to find it...

Comment (tvdijen) [reply lib/SimpleSAML/XML/Parser.php]: The return-value is also an issue.. Psalm will say that 'object' does not match the return-type 'Parser'.
I've added a test for the false-case.. it may happen if the output is an empty string, or if PHP runs out of memory, if I read the PHP source code correctly..

File: lib/SimpleSAML/Metadata/SAMLParser.php (hunk 5 above)
Comment (tvdijen) [reply lib/SimpleSAML/Metadata/SAMLParser.php]: We could add a Webmozart-assertion to the saml2-library to catch that?

File: lib/SimpleSAML/XHTML/Template.php (hunk 3 above)
Comment (tvdijen) [reply lib/SimpleSAML/XHTML/Template.php]: I think not.. We completely separated it last year..

File: lib/SimpleSAML/XML/Parser.php (hunk 6 above)
Comment (tvdijen) [reply lib/SimpleSAML/XML/Parser.php]: I've also learned about this a couple of days ago..

**Comment by tvdijen** [comment]
I think I was able to cleanup all of the forced multi-liners by using @psalm-return phpdocs..
Happy?

File: lib/SimpleSAML/Metadata/SAMLParser.php (hunk 5 above)
Comment (jaimeperez) [reply lib/SimpleSAML/Metadata/SAMLParser.php]: I think it's a good idea!

File: lib/SimpleSAML/XHTML/Template.php (hunk 3 above)
Comment (jaimeperez) [reply lib/SimpleSAML/XHTML/Template.php]: I need to check it out, because I know some third-party modules are already implementing twig templates, and it would be unfortunate for those...

File: lib/SimpleSAML/Locale/Translate.php (hunk 2 above)
Comment (jornane) [reply lib/SimpleSAML/Locale/Translate.php]: Agreed that it's a bug in `getPathValue`, but I think the solution is not good either. I'd rather throw an exception if `null === $dictDir`. Your configuration should not work when you set things to `null` that cannot be null.

Comment (tvdijen) [reply lib/SimpleSAML/Locale/Translate.php]: I don't think I agree on a setting that is so fundamental..
`null` can also mean 'unset' in this case and if such a fundamental setting is unset,, a sane default should fill in.. It's just the that the second param of `getPathValue` isn't the safe default in every case..
In any case, this may not be the best fix, but it is what it is until 2.0 at least..
I guess in a future release a bool-param that specifies whether or not a value is mandatory could fix this..

Comment (jornane) [reply lib/SimpleSAML/Locale/Translate.php]: The `dictionarydir` setting is indeed fundamental, that's why the software should crash if it's overridden to an invalid value. `null` is null, 'unset' is unset and a string is a string. When it's a string, use the string. When it's unset, use the default. When it's `null`, crash with a sensible error message.
